diff --git a/src/common/reed_muller.h b/src/common/reed_muller.h
index d4c22b4..114a0cc 100644
--- a/src/common/reed_muller.h
+++ b/src/common/reed_muller.h
@@ -12,5 +12,6 @@
 
 void reed_muller_encode(uint64_t* cdw, const uint64_t* msg);
 void reed_muller_decode(uint64_t* msg, const uint64_t* cdw);
+void reed_muller_decode_one_block(uint64_t* msg, const uint64_t* cdw);
 
 #endif  // HQC_REED_MULLER_H
diff --git a/src/ref/reed_muller.c b/src/ref/reed_muller.c
index 82760e4..00565c4 100644
--- a/src/ref/reed_muller.c
+++ b/src/ref/reed_muller.c
@@ -211,3 +211,21 @@ void reed_muller_decode(uint64_t *msg, const uint64_t *cdw) {
         message_array[i] = find_peaks(&transform);
     }
 }
+
+/* Decode the first RM block */
+void reed_muller_decode_one_block(uint64_t *msg, const uint64_t *cdw) {
+    uint8_t *message_array = (uint8_t *)msg;
+    rm_codeword_t *codeArray = (rm_codeword_t *)cdw;
+    rm_expanded_cdw expanded;
+    for (size_t i = 0; i < 1; i++) {
+        // collect the codewords
+        expand_and_sum(&expanded, &codeArray[i * MULTIPLICITY]);
+        // apply hadamard transform
+        rm_expanded_cdw transform;
+        hadamard(&expanded, &transform);
+        // fix the first entry to get the half Hadamard transform
+        transform[0] -= 64 * MULTIPLICITY;
+        // finish the decoding
+        message_array[i] = find_peaks(&transform);
+    }
+}
